### ISSUE TITLE
Feature Proposal: Editable Custom Table Attributes

### ISSUE THREAD

**Issue by jkulzick** [amundsen-io/amundsen#1372]
Users of Amundsen occasionally want to display table level attributes on the table detail page that are specific to their business or data source. Unlike programmatic descriptions, they also want to be able to edit this information directly in the UI as these attributes are typically human-generated.

Some examples include:

- retention policy
- data usage policy

## Expected Behavior or Use Case

Users can display and edit additional custom attributes in the table detail page.

## Service or Ingestion ETL

frontend and metadata services

## Possible Implementation

Define additional custom table attributes via configuration in the frontend. The custom table attributes are then displayed in the table detail page using the `EditableText` component. Custom attributes are persisted to the graph using the metadata service and a new PATCH table API endpoint.

## Context

This would allow users to add human-generated business-specific metadata to Amundsen and maintain it directly in the UI.

**Comment by sewardgw** [comment]
Love this idea for more structured content! This feels like it could be straight forward enough that we may be able hash out the details here and it may not need to go through the full RFC process. As part of that I'd love to just understand a little more how you envision this working and what the potential backwards compatibility impact would be:

- Do you want to allow _any_ arbitrary key/values to be stored?
- Do you have a recommendation for how this should be stored in the DB (e.g. new nodes, table attributes)?
- Do you need a similar entry point to creating these objects within data builder?

**Comment by jkulzick** [comment]
Thank you! These are great questions. I've included some thoughts below.

> Do you want to allow _any_ arbitrary key/values to be stored?

Ideally, the metadata service would be aware of what custom attributes are configured so that it only stores those keys. It would probably be good for the metadata service to be aware of the types for these keys as well so it can cast the values just to be safe.

Also, we should only allow basic types, such as string, int, date, etc.

> Do you have a recommendation for how this should be stored in the DB (e.g. new nodes, table attributes)?

We could store the custom attributes as properties on the Table node. A value for a custom attribute would be added via the table details page and therefore would only be associated with that table. A new nodes feels a bit heavy since that node would only have one property and then we would need a `key` property as well.

> Do you need a similar entry point to creating these objects within data builder?

It would be a good idea to have an entry point in data builder to give users a way to seed custom attributes or to programmatically load some or all custom attributes on a re-occurring basis.

For modeling:

One approach would be to incorporate custom attributes into `TableMetadata` by introducing a new named argument of type dict. Then you can unpack that named argument and merge it into `attributes` when instantiating a `GraphNode`.

Another more generic approach would be to add a new databuilder model called something like `CustomMetadata` that could be re-used. It would accept a `key`, `label` and dict of custom attributes to pass to `GraphNode` as `attributes`.

AFAIK the current implementation of `Neo4jCsvPublisher` should set these custom attributes as properties on the node either on create or on match (as configured by the user via `NEO4J_CREATE_ONLY_NODES`). Also, because we define `SET` for each key-value pair in `Neo4jCsvPublisher`, only the provided custom attributes would be affected.

The actual data for custom attributes could be extracted separately from `TableMetadata` or provided in it's own csv file. It could also be extracted alongside `TableMetadata`.

**Comment by sewardgw** [comment]
I think this makes sense and the way you're describing is a pure addition, non-breaking change which feels like this issue is properly scoped within the issue. The one thing I think you may want to consider is the read side for these editable attributes.

One of the biggest benefits of prog descriptions being modeled as a separate node is that it standardizes the location and formatting of prog descriptions. This way, when submitting a query to the graph you can essentially wrap any of the top-level assets with a relationship to the prog desc, for example (sudo coded):
```
(TABLE) -[HAS_PROG_DESC:]-> (PROG_DESC)
#
(FEATURE) -[HAS_PROG_DESC:]-> (PROG_DESC)
# 
(DASHBOARD) -[HAS_PROG_DESC:]-> (PROG_DESC)
```
I think this approach has major benefits in 3 scenarios: 
1. If you dynamically building queries (which we're not doing _right now_) 
2. it makes it easier to bulk adjust programatic descriptions if the format ever changes without having to write a query that impacts all of the dashboards, features, tables, etc. that would otherwise have them embedded within their individual nodes.
3. If you want to reuse the same prog desc across assets

**Comment by feng-tao** [comment]
if it is human generated, qqs: 
1. why not leverage table descriptions? 
2. if table description not the case, how about tags? 
3. who could add description custom attributes and how to maintain it?

The feature seems to be overlap with a few existing features which we need to discuss in more details (RFC is required)

**Comment by feng-tao** [comment]
for the two examples you listed: retention policy or data usage policy,  the table badge with different category seems to be a better leverage.

**Comment by markgrover** [comment]
Hey @feng-tao,

> why not leverage table descriptions?
Table descriptions are meant for descriptions. Even at Lyft, we had so many requests from the DE team requesting things like granularity and governance policy as separate fields. I don't think it's right to club up under descriptions, it's a hack at best. This brings 1st class citizen support for those additional fields.

> if table description not the case, how about tags?
Tags are the wrong construct here as well, IMO. This is because tags are limited while here for the a field like security policy, we want to be able to attach a hyperlink that has the policy. The tag isn't the right construct.

> who could add description custom attributes and how to maintain it?
Same thing as the description field.

In general, I feel like this has been a limitation of Amundsen for a long time, and it's been a long time coming. I don't have a strong opinion on RFC but my meta feedback here is that our dev velocity as a project is really slow so if we can move this forward rather fast, that'd be great for the project and the community.

Let me know if you have any further questions or concerns here.

**Comment by feng-tao** [comment]
> Hey @feng-tao,
> 
> > why not leverage table descriptions?
> > Table descriptions are meant for descriptions. Even at Lyft, we had so many requests from the DE team requesting things like granularity and governance policy as separate fields. I don't think it's right to club up under descriptions, it's a hack at best. This brings 1st class citizen support for those additional fields.
> 
> > if table description not the case, how about tags?
> > Tags are the wrong construct here as well, IMO. This is because tags are limited while here for the a field like security policy, we want to be able to attach a hyperlink that has the policy. The tag isn't the right construct.
> 
> > who could add description custom attributes and how to maintain it?
> > Same thing as the description field.
> 
> In general, I feel like this has been a limitation of Amundsen for a long time, and it's been a long time coming. I don't have a strong opinion on RFC but my meta feedback here is that our dev velocity as a project is really slow so if we can move this forward rather fast, that'd be great for the project and the community.
> 
> Let me know if you have any further questions or concerns here.

A few comments:
1. what will the UX look like (e.g what are the total custom attributes to support? how it vs prog desc features? who could update ?)? cc @Golodhros 
2. Amundsen supports multiple data store. what metadata store does this feature support?
3. Based on my experiences, lots of community members have more concern on UI editing than support the actual editing as it doesn't have ACL model. People tend to disable descriptions for most places. 
4. what are the permissions models like who could update/edit/maintain?

In short, given this is an engineering feature, I do think we need proper discussion which is the reason RFC exists that engineers who build and maintain the project could agree upon the direction.

cc @Golodhros @allisonsuarez @dkunitsk @mgorsk1 @youngyjd

**Comment by markgrover** [comment]
It's John's PR so he's in a better place to answer these. Fine to do RFC.

BTW, I disagree that "most places disable descriptions" but that's beside the point. The permissions model here will be exactly the same as the permissions model for descriptions.

**Comment by feng-tao** [comment]
> It's John's PR so he's in a better place to answer these. Fine to do RFC.
> 
> BTW, I disagree that "most places disable descriptions" but that's beside the point. The permissions model here will be exactly the same as the permissions model for descriptions.

so what exactly is the permission model then? If we allow anyone add/update any custom attribute without review, how do we handle the following scenarios like: 1. I mark this table as retention 7 years, someone I didn't know change it to 30 days; 2 overlap attributes with typo(someone put `retention` or `RetentionTIME`); 3. what if we define the table attribute somewhere which could be ingested in prog description but other unknown person add similar table attribute through UI.

Lots of user experiences about adding custom arbitrary attributes are unknown.

yeah, sure :) I don't think we need to agree upon everything:)  That's why a RFC could help to understand the details, use case and product UX and the engineering details .

**Comment by markgrover** [comment]
Good questions. Those can be handled in the rfc. Thanks Tao.

**Comment by mgorsk1** [comment]
+1 on RFC, and +1 definitely could use having more ways to create custom attributes in the UI. There are quite a few things this proposal poked (permissions model, purpose of other UI generated metadata, overlap with prog desc) and it seems like a good opportunity to hash some of those out. We are pretty quick on RFC so shouldn't be a significant delay.

In the end we even might end up with something as simple as 'add possibility to define prog descriptions in the UI' (which would essentially make the name programmatic description obsolete)

**Comment by jkulzick** [comment]
Thank you everyone for the great discussion. A lot of good comments and questions. I will re-review the aforementioned existing features for this use case (badges, programmatic descriptions,  etc.) and then work on putting together an RFC.

**Comment by jkulzick** [comment]
@mgorsk1 @feng-tao @sewardgw @markgrover @dorianj 

After taking another look at existing Amundsen features (tags, badges, table descriptions, programmatic descriptions), I would like to change my feature proposal to "Editable Programmatic Descriptions". Given that, I wanted to confirm what the best next steps would be to move this proposal forward? 1) submit an RFC for Editable Programmatic Descriptions 2) update this feature request or open a new feature request to document the revised feature request 3) other

Here is some context behind my update:

- The values we will store will include longer text, hyperlinks, etc. For that reason, tags does not seem like a good use based on my understanding of the feature.
- Overloading table descriptions with additional metadata/attributes does not provide a good UX IMHO. Users need to be able to refer to a dedicated component in the UI.
- There was already some [discussion](https://github.com/amundsen-io/amundsen/issues/147#issuecomment-566302027) in the original programmatic descriptions feature proposal around allowing a source to be flagged as editable.
- The permission model for editing editable programmatic descriptions would be the same as table descriptions.

**Comment by markgrover** [comment]
I like the direction.

My vote is #2.

**Comment by feng-tao** [comment]
we should follow the process (similar request is done for https://github.com/amundsen-io/amundsen/pull/1073). Let's do a quick RFC for the process.

**Comment by feng-tao** [comment]
But I think support prog description makes sense with flag.

**Comment by mgorsk1** [comment]
I like idea of editing programmatic descriptions from UI as well. I suspected that this might eventually end up like that.

It would touch changing UI so quick RFC with mocks could be useful but I think we are all on the same page re desirability ;)

**Comment by jkulzick** [comment]
Thanks everyone. I will put together a quick RFC.

**Comment by jkulzick** [comment]
Linking the [RFC](https://github.com/amundsen-io/rfcs/pull/42)
